### PR TITLE
Implement a Flux store.

### PR DESCRIPTION
**Pull request by n1k0**
Ready for feedback.

### CODE REVIEW

**Comment by n1k0** [comment]
Tests are failing because this is WiP.

File: src/js/flux.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  getInitialState: function() {
+    return {
+      articles: [],
+      current: null,
+      error: null
+    };
+  },
+
+  _wrap: function(promise) {
+    return promise.catch(err => {
```
Comment (leplatrem) [inline src/js/flux.js]: OMG, this is the new `bind(this)` !

File: src/js/app.js [hunk 2]
```diff
@@ -1,24 +1,109 @@
...
 
-var React = require("react");
 // Load 6to5 polyfills http://6to5.org/docs/usage/polyfill/
 require("6to5/polyfill");
 
-var App = React.createClass({
+var React = require("react/addons");
+var DocBrown = require("docbrown");
+
+var {ArticleActions, ArticleStore} = require("./flux");
```
Comment (leplatrem) [inline src/js/app.js]: Neat! _(first time I see this :))_

Comment (Natim) [inline src/js/app.js]: Hopefully we had @bnjbvr conf to be able to understand.

**Comment by n1k0** [comment]
@leplatrem @ametaireau @Natim I know this is pretty huge, though would you mind giving it a (quick) look?

File: README.md [hunk 3]
```diff
@@ -15,7 +15,9 @@ Required nodejs v0.10.29+ and npm v2.1.0+.
 Configuration
 -------------
 
-TODO
+You can configure the client using the following environment variables:
+
+- `READINGLIST_SERVER_BASEURL`: the Readinglist server base URL; default: `http://0.0.0.0:8000/v0`
```
Comment (Natim) [inline README.md:20]: Shouldn't we use a file to configure that? `config.js` with some configuration?

File: src/js/api.js [hunk 4]
```diff
@@ -71,7 +81,14 @@ class API {
    * @return {Promise}
    */
   listArticles(params={}) {
-    return this._wrap(this.client({path: "/articles"}));
+    var q = {
+      unread:   true,
+      _sort: "-last_modified,title"
```
Comment (Natim) [inline src/js/api.js]: `last_modified` is unique so you can remove `title`.

File: src/js/api.js [hunk 5]
```diff
@@ -91,9 +108,9 @@ class API {
   deleteArticle(params={}) {
     return this._wrap(this.client({
       method: "DELETE",
-      path: `/article/${params.id}`
+      path: `/articles/${params._id}`
```
Comment (Natim) [inline src/js/api.js]: https://github.com/mozilla-services/readinglist/pull/91 is going to change that I think.

File: README.md (hunk 3 above)
Comment (n1k0) [inline README.md:20]: In the future we'll use [envify](https://github.com/hughsk/envify), which will compile the js assets and alter them according to that env var. That way, deployment is mostly matter of building the project using the right vars set.

File: src/js/main.js [hunk 6]
```diff
@@ -1,10 +1,15 @@
...
-var API = require("./api.js");
-var baseUrl = process.env.READINGLIST_SERVER_BASEURL || "http://0.0.0.0:8000/v0";
+import React from "react/addons";
+import API from "./api";
+import App from "./app";
+import {ArticleStore, stores} from "./flux";
 
-var api = new API(baseUrl);
+// XXX handle this with envify
+var serverUrl = process.env.READINGLIST_SERVER_BASEURL || "http://0.0.0.0:8000/v0";
```
Comment (Natim) [inline src/js/main.js:9]: What is the ENV of a webpage?

File: src/js/api.js (hunk 5 above)
Comment (n1k0) [inline src/js/api.js]: Yeah, whatever commit lands first :p

File: src/js/main.js (hunk 6 above)
Comment (n1k0) [inline src/js/main.js:9]: This `process.env` thing is provided by browserify, to emulate a real nodejs environment. As mentioned previously, we'll use envify to transform these at compilation time.

Comment (Natim) [inline src/js/main.js:9]: Great :)

File: README.md (hunk 3 above)
Comment (almet) [inline README.md:20]: The problem I see is that env vars don't exist on static hosting, and that's how I expect this to be hosted.

File: src/js/api.js [hunk 7]
```diff
@@ -1,21 +1,21 @@
 "use strict";
 
-var rest = require("rest");
-var pathPrefix = require("rest/interceptor/pathPrefix");
-var mime = require("rest/interceptor/mime");
-var errorCode = require("rest/interceptor/errorCode");
-var defaultRequest = require("rest/interceptor/defaultRequest");
-var hateoas = require("rest/interceptor/hateoas");
+import rest from "rest";
+import pathPrefix from "rest/interceptor/pathPrefix";
```
Comment (almet) [inline src/js/api.js:4]: hmm, import from :)

File: src/js/api.js [hunk 8]
```diff
@@ -1,21 +1,21 @@
...
+import mime from "rest/interceptor/mime";
+import errorCode from "rest/interceptor/errorCode";
+import defaultRequest from "rest/interceptor/defaultRequest";
+import hateoas from "rest/interceptor/hateoas";
```
Comment (almet) [inline src/js/api.js:8]: What is this used for?

Comment (almet) [inline src/js/api.js:8]: I don't think the readinglist serverr is using HATEOS

File: src/js/api.js [hunk 9]
```diff
@@ -24,17 +24,31 @@ class API {
...
-    //.done(function(res) {
-    //   return res.entity;
-    // })
+  }
+
+  createClient() {
+    return rest
+      .wrap(pathPrefix, {prefix: this.baseUrl})
+      .wrap(errorCode, {code: 400})
+      .wrap(defaultRequest, {headers: {"X-Requested-With": "readinglist-client"}})
```
Comment (almet) [inline src/js/api.js:47]: The  `X` header had been deprecated. See RFC 6648 — http://tools.ietf.org/rfc/rfc6648.txt

Comment (Natim) [inline src/js/api.js:47]: I think we can directly use `Requested-With`

File: src/js/components/ArticleEntry.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  },
+
+  render: function() {
+    return (
+      <div>
+        <h4>
+          <a href={this.props.article.resolved_url}>
+            {this.props.article.resolved_title}
+          </a>
+          {" "}
```
Comment (almet) [inline src/js/components/ArticleEntry.js:20]: Why are we doing this?

File: src/js/components/ArticleForm.js [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    return (
+      <Panel title="Add an article">
+        <form method="post" onSubmit={this.handleSubmit}>
+          <div className="form-group">
+            <input className="form-control" type="text"
+                    placeholder="Title" valueLink={this.linkState("title")}/>
+          </div>
+          <div className="form-group">
+            <input className="form-control" type="url"
+                    placeholder="http://" valueLink={this.linkState("url")}/>
```
Comment (almet) [inline src/js/components/ArticleForm.js:35]: Should we use  `///` here rather than http?

File: src/js/components/ArticleList.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    return JSON.stringify(this.props) !== JSON.stringify(nextProps);
+  },
+
+  render: function() {
+    return (
+      <Panel title="Articles" bodyWrap={false}>
+        <ul className="list-group">{
+          this.props.articles.map(function(article, i) {
+            return (
+              <li key={i} className="list-group-item">
```
Comment (almet) [inline src/js/components/ArticleList.js:20]: If articles already have an id, should we use it as the key?

File: src/js/components/ArticleList.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+                <ArticleEntry article={article} />
+              </li>
+            );
+          })
+        }</ul>
+        {this.props.articles.length === 0 ?
+          <p className="list-group-item">Your list is empty.</p> : null}
```
Comment (almet) [inline src/js/components/ArticleList.js:27]: nit: should we change that to say "You don"t have anything to read just yet" ?

File: src/js/flux.js [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+  listSuccess: function(articles) {
+    this.setState({articles: articles});
+  },
+
+  listError: function(err) {
+    this.setError(err, "list");
+  },
+});
+
+function StoreRegistry() {
```
Comment (almet) [inline src/js/flux.js:106]: Should this be a part of docbrown directly?

File: src/js/components/ArticleForm.js (hunk 11 above)
Comment (n1k0) [inline src/js/components/ArticleForm.js:35]: That's just a placeholder to indicate we expect a url for that field :)

File: src/js/components/ArticleList.js (hunk 12 above)
Comment (n1k0) [inline src/js/components/ArticleList.js:20]: Good idea!

File: src/js/components/ArticleList.js (hunk 13 above)
Comment (n1k0) [inline src/js/components/ArticleList.js:27]: Sold!

File: src/js/flux.js (hunk 14 above)
Comment (n1k0) [inline src/js/flux.js:106]: I don't think so, as this is really ArticleStore implementation detail.

File: src/js/api.js (hunk 8 above)
Comment (n1k0) [inline src/js/api.js:8]: This is used to automatically convert JSON bodies to entities (for now). I wish we'll use hateoas someday anyway :)

File: src/js/api.js (hunk 9 above)
Comment (n1k0) [inline src/js/api.js:47]: Good point, will fix.

File: src/js/components/ArticleEntry.js (hunk 10 above)
Comment (n1k0) [inline src/js/components/ArticleEntry.js:20]: That's the only way I know to add whitespace between tags in JSX.

File: README.md (hunk 3 above)
Comment (n1k0) [inline README.md:20]: The deployment process will be:
- fetch the sources from the repo
- define config env var, eg. server base url as documented here
- build the dist files for the client app
- that build will reflect the configured base url using the env var

That simple :)

Comment (n1k0) [inline README.md:20]: To be clear, I mean we could totally generate the dist files before even sending them to that static hosting solution.

File: src/js/api.js (hunk 8 above)
Comment (almet) [inline src/js/api.js:8]: Then we should probably start the discussion :)

File: src/js/components/ArticleForm.js (hunk 11 above)
Comment (almet) [inline src/js/components/ArticleForm.js:35]: Ah, well...

**Pull request merged**